File: swift/Tests/KiahkTests/KiahkTests.swift
import XCTest
import Foundation
@testable import Kiahk

// ----------------------------------------------------------------------
// Shared test-vector loader.
//
// `#file` resolves at compile time to this source file's path. From
// swift/Tests/KiahkTests/KiahkTests.swift we walk up three levels to
// reach the repo root, then into core/.
// ----------------------------------------------------------------------

private let coreDir: URL = {
    let thisFile = URL(fileURLWithPath: #file)
    return thisFile
        .deletingLastPathComponent() // KiahkTests
        .deletingLastPathComponent() // Tests
        .deletingLastPathComponent() // swift
        .deletingLastPathComponent() // repo root
        .appendingPathComponent("core")
}()

private struct GregCoptic: Decodable {
    let gregorian: YMD
    let coptic: YMD
}
private struct YMD: Decodable {
    let year: Int
    let month: Int
    let day: Int
}
private struct EasterVec: Decodable {
    let gregorian_year: Int
    let date: YMD
}
private struct MoveableVec: Decodable {
    let gregorian_year: Int
    let feast_id: String
    let date: YMD
}
private struct InvalidDate: Decodable {
    let year: Int
    let month: Int
    let day: Int
}
private struct Vectors: Decodable {
    let gregorian_to_coptic: [GregCoptic]
    let coptic_to_gregorian: [GregCoptic]
    let easter: [EasterVec]
    let moveable_feasts: [MoveableVec]
    let invalid_coptic_dates: [InvalidDate]
    let invalid_gregorian_dates: [InvalidDate]
}

private let vectors: Vectors = {
    let url = coreDir.appendingPathComponent("test-vectors.json")
    let data = try! Data(contentsOf: url)
    return try! JSONDecoder().decode(Vectors.self, from: data)
}()

// ----------------------------------------------------------------------
// Error cases
// ----------------------------------------------------------------------

final class ErrorsTests: XCTestCase {
    func testInvalidCopticDateCaseExists() {
        let err = KiahkError.invalidCopticDate(year: 1741, month: 14, day: 1, reason: "month out of range")
        if case .invalidCopticDate = err {} else { XCTFail("expected .invalidCopticDate, got \(err)") }
    }

    func testInvalidGregorianDateCaseExists() {
        let err = KiahkError.invalidGregorianDate(year: 2025, month: 2, day: 29, reason: "not a leap year")
        if case .invalidGregorianDate = err {} else { XCTFail("expected .invalidGregorianDate, got \(err)") }
    }

    func testUnsupportedLocaleCaseExists() {
        let err = KiahkError.unsupportedLocale(feastID: "easter", locale: "fr")
        if case .unsupportedLocale = err {} else { XCTFail("expected .unsupportedLocale, got \(err)") }
    }

    func testErrorTypeConformsToErrorProtocol() {
        let err: Error = KiahkError.invalidCopticDate(year: 0, month: 0, day: 0, reason: "")
        XCTAssertNotNil(err)
    }
}

// ----------------------------------------------------------------------
// algorithms / gregorianToJdn
// ----------------------------------------------------------------------

final class GregorianToJdnTests: XCTestCase {
    func testKnownValues() {
        XCTAssertEqual(gregorianToJdn(year: 2000, month: 1, day: 1), 2451545)
        XCTAssertEqual(gregorianToJdn(year: 1900, month: 1, day: 1), 2415021)
        XCTAssertEqual(gregorianToJdn(year: 2025, month: 1, day: 11), 2460687)
    }
}

final class JdnToGregorianTests: XCTestCase {
    func testRoundTrip() {
        let cases: [(Int, Int, Int)] = [
            (2000, 1, 1), (1900, 1, 1), (2025, 1, 11), (2024, 12, 25), (2025, 9, 11),
        ]
        for (y, m, d) in cases {
            let jdn = gregorianToJdn(year: y, month: m, day: d)
            let r = jdnToGregorian(jdn)
            XCTAssertEqual(r.year, y, "round trip year for \(y)-\(m)-\(d)")
            XCTAssertEqual(r.month, m, "round trip month for \(y)-\(m)-\(d)")
            XCTAssertEqual(r.day, d, "round trip day for \(y)-\(m)-\(d)")
        }
    }
}

final class CopticToJdnTests: XCTestCase {
    func testEpoch() {
        // 1 Tout 1 AM is JDN 1_825_030 (the Coptic epoch).
        XCTAssertEqual(copticToJdn(year: 1, month: 1, day: 1), 1_825_030)
    }
}

final class JdnToCopticTests: XCTestCase {
    func testCopticRoundTripVectors() {
        for vec in vectors.gregorian_to_coptic {
            let c = vec.coptic
            let jdn = copticToJdn(year: c.year, month: c.month, day: c.day)
            let r = jdnToCoptic(jdn)
            XCTAssertEqual(r.year, c.year, "year for \(c.year)/\(c.month)/\(c.day)")
            XCTAssertEqual(r.month, c.month, "month")
            XCTAssertEqual(r.day, c.day, "day")
        }
    }
}

final class GregorianToCopticTests: XCTestCase {
    func testVectors() {
        for vec in vectors.gregorian_to_coptic {
            let g = vec.gregorian
            let c = vec.coptic
            let r = gregorianToCoptic(year: g.year, month: g.month, day: g.day)
            XCTAssertEqual(r.year, c.year, "G \(g.year)-\(g.month)-\(g.day) → C year")
            XCTAssertEqual(r.month, c.month, "G \(g.year)-\(g.month)-\(g.day) → C month")
            XCTAssertEqual(r.day, c.day, "G \(g.year)-\(g.month)-\(g.day) → C day")
        }
    }
}

final class CopticToGregorianTests: XCTestCase {
    func testVectors() {
        for vec in vectors.coptic_to_gregorian {
            let c = vec.coptic
            let g = vec.gregorian
            let r = copticToGregorian(year: c.year, month: c.month, day: c.day)
            XCTAssertEqual(r.year, g.year, "C \(c.year)/\(c.month)/\(c.day) → G year")
            XCTAssertEqual(r.month, g.month)
            XCTAssertEqual(r.day, g.day)
        }
    }
}

final class ComputeEasterTests: XCTestCase {
    func testVectors() {
        for vec in vectors.easter {
            let r = computeEaster(gregorianYear: vec.gregorian_year)
            XCTAssertEqual(r.year, vec.date.year, "easter \(vec.gregorian_year) year")
            XCTAssertEqual(r.month, vec.date.month, "easter \(vec.gregorian_year) month")
            XCTAssertEqual(r.day, vec.date.day, "easter \(vec.gregorian_year) day")
        }
    }
}

final class AddDaysTests: XCTestCase {
    func testKnownOffsets() {
        var r = addDays(year: 2025, month: 1, day: 1, days: 10)
        XCTAssertEqual(r.year, 2025); XCTAssertEqual(r.month, 1); XCTAssertEqual(r.day, 11)

        r = addDays(year: 2025, month: 1, day: 1, days: -1)
        XCTAssertEqual(r.year, 2024); XCTAssertEqual(r.month, 12); XCTAssertEqual(r.day, 31)

        // 2024 is a leap year
        r = addDays(year: 2024, month: 2, day: 28, days: 1)
        XCTAssertEqual(r.year, 2024); XCTAssertEqual(r.month, 2); XCTAssertEqual(r.day, 29)
    }
}

// ----------------------------------------------------------------------
// GregorianDate
// ----------------------------------------------------------------------

final class GregorianDateTests: XCTestCase {
    func testBasicConstruction() throws {
        let g = try GregorianDate(year: 2025, month: 1, day: 11)
        XCTAssertEqual(g.year, 2025)
        XCTAssertEqual(g.month, 1)
        XCTAssertEqual(g.day, 11)
    }

    func testRejectsInvalidDates() {
        for bad in vectors.invalid_gregorian_dates {
            XCTAssertThrowsError(try GregorianDate(year: bad.year, month: bad.month, day: bad.day)) { err in
                guard case KiahkError.invalidGregorianDate = err else {
                    XCTFail("expected .invalidGregorianDate for \(bad), got \(err)")
                    return
                }
            }
        }
    }

    func testToDate() throws {
        let g = try GregorianDate(year: 2025, month: 1, day: 11)
        let date = g.toDate()
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = TimeZone(identifier: "UTC")!
        let comps = cal.dateComponents([.year, .month, .day], from: date)
        XCTAssertEqual(comps.year, 2025)
        XCTAssertEqual(comps.month, 1)
        XCTAssertEqual(comps.day, 11)
    }

    func testFromDate() {
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = TimeZone(identifier: "UTC")!
        let date = cal.date(from: DateComponents(year: 2025, month: 1, day: 11, hour: 12))!
        let g = GregorianDate(date: date)
        XCTAssertEqual(g.year, 2025)
        XCTAssertEqual(g.month, 1)
        XCTAssertEqual(g.day, 11)
    }

    func testEquatable() throws {
        let a = try GregorianDate(year: 2025, month: 1, day: 11)
        let b = try GregorianDate(year: 2025, month: 1, day: 11)
        XCTAssertEqual(a, b)
        XCTAssertEqual(a.hashValue, b.hashValue)
    }
}

// ----------------------------------------------------------------------
// CopticDate
// ----------------------------------------------------------------------

final class CopticDateTests: XCTestCase {
    func testBasicConstruction() throws {
        let c = try CopticDate(year: 1741, month: 5, day: 3)
        XCTAssertEqual(c.year, 1741)
        XCTAssertEqual(c.month, 5)
        XCTAssertEqual(c.day, 3)
    }

    func testRejectsInvalidDates() {
        for bad in vectors.invalid_coptic_dates {
            XCTAssertThrowsError(try CopticDate(year: bad.year, month: bad.month, day: bad.day)) { err in
                guard case KiahkError.invalidCopticDate = err else {
                    XCTFail("expected .invalidCopticDate for \(bad), got \(err)")
                    return
                }
            }
        }
    }

    func testToGregorian() throws {
        let c = try CopticDate(year: 1741, month: 5, day: 3)
        let g = try c.toGregorian()
        XCTAssertEqual(g.year, 2025)
        XCTAssertEqual(g.month, 1)
        XCTAssertEqual(g.day, 11)
    }

    func testGregorianToCopticInstance() throws {
        let g = try GregorianDate(year: 2025, month: 1, day: 11)
        let c = try g.toCoptic()
        XCTAssertEqual(c.year, 1741)
        XCTAssertEqual(c.month, 5)
        XCTAssertEqual(c.day, 3)
    }
}

// ----------------------------------------------------------------------
// feasts data parity with core/feasts.json
// ----------------------------------------------------------------------

final class FeastsDataTests: XCTestCase {
    func testMatchesCoreFeastsJSON() throws {
        let url = coreDir.appendingPathComponent("feasts.json")
        let data = try Data(contentsOf: url)
        let core = try JSONSerialization.jsonObject(with: data) as! [[String: Any]]

        XCTAssertEqual(kFeasts.count, core.count)
        for (i, f) in kFeasts.enumerated() {
            let ref = core[i]
            XCTAssertEqual(f.id, ref["id"] as? String, "id [\(i)]")
            XCTAssertEqual(f.type, ref["type"] as? String, "type [\(i)]")
            XCTAssertEqual(f.category, ref["category"] as? String, "category [\(i)]")
            let refNames = ref["names"] as! [String: String]
            XCTAssertEqual(f.names["en"], refNames["en"], "en [\(i)]")
            XCTAssertEqual(f.names["ar"], refNames["ar"], "ar [\(i)]")
        }
    }
}

// ----------------------------------------------------------------------
// Feast struct + name(locale:)
// ----------------------------------------------------------------------

final class FeastTests: XCTestCase {
    func testBasicFieldsAndName() throws {
        let g = try GregorianDate(year: 2025, month: 4, day: 20)
        let rec = try feastByID("easter")
        let feast = Feast(
            id: rec.id, type: rec.type, category: rec.category,
            names: rec.names, gregorianDate: g
        )
        XCTAssertEqual(feast.id, "easter")
        XCTAssertEqual(feast.type, "moveable")
        XCTAssertEqual(feast.category, "major")
        XCTAssertEqual(try feast.name(locale: "en"), "Easter Sunday")
        XCTAssertEqual(try feast.name(locale: "ar"), "عيد القيامة المجيد")
    }

    func testUnknownLocaleThrows() throws {
        let g = try GregorianDate(year: 2025, month: 4, day: 20)
        let rec = try feastByID("easter")
        let feast = Feast(
            id: rec.id, type: rec.type, category: rec.category,
            names: rec.names, gregorianDate: g
        )
        XCTAssertThrowsError(try feast.name(locale: "fr")) { err in
            guard case KiahkError.unsupportedLocale = err else {
                XCTFail("expected .unsupportedLocale, got \(err)")
                return
            }
        }
    }
}

// ----------------------------------------------------------------------
// CopticCalendar.easterDate
// ----------------------------------------------------------------------

final class CopticCalendarEasterDateTests: XCTestCase {
    func testVectors() throws {
        for vec in vectors.easter {
            let g = try CopticCalendar.easterDate(gregorianYear: vec.gregorian_year)
            XCTAssertEqual(g.year, vec.date.year, "easter \(vec.gregorian_year) year")
            XCTAssertEqual(g.month, vec.date.month, "easter \(vec.gregorian_year) month")
            XCTAssertEqual(g.day, vec.date.day, "easter \(vec.gregorian_year) day")
        }
    }
}

// ----------------------------------------------------------------------
// CopticCalendar.moveableFeast
// ----------------------------------------------------------------------

final class CopticCalendarMoveableFeastTests: XCTestCase {
    func testVectors() throws {
        for vec in vectors.moveable_feasts {
            let feast = try CopticCalendar.moveableFeast(id: vec.feast_id, gregorianYear: vec.gregorian_year)
            XCTAssertEqual(feast.id, vec.feast_id)
            let g = feast.gregorianDate
            XCTAssertEqual(g.year, vec.date.year, "\(vec.feast_id) \(vec.gregorian_year) year")
            XCTAssertEqual(g.month, vec.date.month)
            XCTAssertEqual(g.day, vec.date.day)
        }
    }
}

// ----------------------------------------------------------------------
// CopticCalendar.yearFeasts
// ----------------------------------------------------------------------

final class CopticCalendarYearFeastsTests: XCTestCase {
    func testNonEmptyAndSorted() {
        let feasts = CopticCalendar.yearFeasts(gregorianYear: 2025)
        XCTAssertFalse(feasts.isEmpty)
        for i in 1..<feasts.count {
            let a = feasts[i - 1].gregorianDate
            let b = feasts[i].gregorianDate
            XCTAssertTrue(_lessOrEqual(a, b), "feasts not sorted at \(i): \(a) > \(b)")
        }
    }

    func testIncludesEaster() {
        let feasts = CopticCalendar.yearFeasts(gregorianYear: 2025)
        guard let easter = feasts.first(where: { $0.id == "easter" }) else {
            XCTFail("Easter not in yearFeasts(2025)")
            return
        }
        XCTAssertEqual(easter.gregorianDate.year, 2025)
        XCTAssertEqual(easter.gregorianDate.month, 4)
        XCTAssertEqual(easter.gregorianDate.day, 20)
    }

    private func _lessOrEqual(_ a: GregorianDate, _ b: GregorianDate) -> Bool {
        if a.year != b.year { return a.year < b.year }
        if a.month != b.month { return a.month < b.month }
        return a.day <= b.day
    }
}

File: csharp/Kiahk.Tests/KiahkTests.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text.Json;
using System.Text.Json.Serialization;
using Kiahk;
using Xunit;

namespace Kiahk.Tests;

// ----------------------------------------------------------------------
// Shared test-vector loader.
//
// [CallerFilePath] resolves at compile time to this source file's path.
// From csharp/Kiahk.Tests/KiahkTests.cs we walk up two levels to reach
// the repo root, then into core/.
// ----------------------------------------------------------------------

internal static class Paths
{
    private static string CallerDir([CallerFilePath] string callerPath = "")
        => new FileInfo(callerPath).Directory!.FullName;

    public static string CoreDir
    {
        get
        {
            // CallerDir() → .../csharp/Kiahk.Tests
            var testsDir = new DirectoryInfo(CallerDir());
            return Path.Combine(testsDir.Parent!.Parent!.FullName, "core");
        }
    }
}

internal sealed record YMD(int Year, int Month, int Day);
internal sealed record GregCoptic(YMD Gregorian, YMD Coptic);
internal sealed record EasterVec(int Gregorian_Year, YMD Date);
internal sealed record MoveableVec(int Gregorian_Year, string Feast_Id, YMD Date);
internal sealed record MonthNameVec(int Month, string Locale, string Name);
internal sealed record MonthLocaleVec(int Month, string Locale);
internal sealed record Vectors(
    GregCoptic[] Gregorian_To_Coptic,
    GregCoptic[] Coptic_To_Gregorian,
    EasterVec[] Easter,
    MoveableVec[] Moveable_Feasts,
    YMD[] Invalid_Coptic_Dates,
    YMD[] Invalid_Gregorian_Dates,
    MonthNameVec[] Coptic_Month_Names,
    MonthLocaleVec[] Invalid_Coptic_Month_Locales,
    int[] Invalid_Coptic_Months_For_Name);

internal static class TestVectors
{
    public static readonly Vectors V = LoadVectors();

    private static Vectors LoadVectors()
    {
        var path = Path.Combine(Paths.CoreDir, "test-vectors.json");
        var json = File.ReadAllText(path);
        var opts = new JsonSerializerOptions
        {
            PropertyNameCaseInsensitive = true,
        };
        return JsonSerializer.Deserialize<Vectors>(json, opts)!;
    }
}

// ----------------------------------------------------------------------
// Error classes
// ----------------------------------------------------------------------

public class ErrorsTests
{
    [Fact]
    public void InvalidCopticDateExceptionExtendsException()
    {
        var ex = new InvalidCopticDateException("bad");
        Assert.IsAssignableFrom<Exception>(ex);
        Assert.Equal("bad", ex.Message);
    }

    [Fact]
    public void InvalidGregorianDateExceptionExtendsException()
    {
        var ex = new InvalidGregorianDateException("bad");
        Assert.IsAssignableFrom<Exception>(ex);
    }

    [Fact]
    public void UnsupportedLocaleExceptionExtendsException()
    {
        var ex = new UnsupportedLocaleException("easter", "fr");
        Assert.IsAssignableFrom<Exception>(ex);
        Assert.Contains("easter", ex.Message);
        Assert.Contains("fr", ex.Message);
    }
}

// ----------------------------------------------------------------------
// Algorithms.GregorianToJdn
// ----------------------------------------------------------------------

public class GregorianToJdnTests
{
    [Theory]
    [InlineData(2000, 1, 1, 2451545)]
    [InlineData(1900, 1, 1, 2415021)]
    [InlineData(2025, 1, 11, 2460687)]
    public void KnownValues(int y, int m, int d, int expectedJdn)
    {
        Assert.Equal(expectedJdn, Algorithms.GregorianToJdn(y, m, d));
    }
}

public class JdnToGregorianTests
{
    public static readonly object[][] RoundTripData =
    {
        new object[] { 2000, 1, 1 },
        new object[] { 1900, 1, 1 },
        new object[] { 2025, 1, 11 },
        new object[] { 2024, 12, 25 },
        new object[] { 2025, 9, 11 },
    };

    [Theory]
    [MemberData(nameof(RoundTripData))]
    public void RoundTrip(int y, int m, int d)
    {
        int jdn = Algorithms.GregorianToJdn(y, m, d);
        var r = Algorithms.JdnToGregorian(jdn);
        Assert.Equal((y, m, d), (r.Year, r.Month, r.Day));
    }
}

public class CopticToJdnTests
{
    [Fact]
    public void EpochIs1825030()
    {
        Assert.Equal(1_825_030, Algorithms.CopticToJdn(1, 1, 1));
    }
}

public class JdnToCopticTests
{
    public static IEnumerable<object[]> CopticVectors() =>
        TestVectors.V.Gregorian_To_Coptic.Select(v => new object[] { v.Coptic.Year, v.Coptic.Month, v.Coptic.Day });

    [Theory]
    [MemberData(nameof(CopticVectors))]
    public void CopticRoundTrip(int cy, int cm, int cd)
    {
        int jdn = Algorithms.CopticToJdn(cy, cm, cd);
        var r = Algorithms.JdnToCoptic(jdn);
        Assert.Equal((cy, cm, cd), (r.Year, r.Month, r.Day));
    }
}

public class GregorianToCopticTests
{
    public static IEnumerable<object[]> Vectors() =>
        TestVectors.V.Gregorian_To_Coptic.Select(v => new object[]
        {
            v.Gregorian.Year, v.Gregorian.Month, v.Gregorian.Day,
            v.Coptic.Year, v.Coptic.Month, v.Coptic.Day,
        });

    [Theory]
    [MemberData(nameof(Vectors))]
    public void VectorMatch(int gy, int gm, int gd, int cy, int cm, int cd)
    {
        var r = Algorithms.GregorianToCoptic(gy, gm, gd);
        Assert.Equal((cy, cm, cd), (r.Year, r.Month, r.Day));
    }
}

public class CopticToGregorianTests
{
    public static IEnumerable<object[]> Vectors() =>
        TestVectors.V.Coptic_To_Gregorian.Select(v => new object[]
        {
            v.Coptic.Year, v.Coptic.Month, v.Coptic.Day,
            v.Gregorian.Year, v.Gregorian.Month, v.Gregorian.Day,
        });

    [Theory]
    [MemberData(nameof(Vectors))]
    public void VectorMatch(int cy, int cm, int cd, int gy, int gm, int gd)
    {
        var r = Algorithms.CopticToGregorian(cy, cm, cd);
        Assert.Equal((gy, gm, gd), (r.Year, r.Month, r.Day));
    }
}

public class ComputeEasterTests
{
    public static IEnumerable<object[]> Vectors() =>
        TestVectors.V.Easter.Select(v => new object[]
        {
            v.Gregorian_Year, v.Date.Year, v.Date.Month, v.Date.Day,
        });

    [Theory]
    [MemberData(nameof(Vectors))]
    public void EasterMatchesVectors(int year, int ey, int em, int ed)
    {
        var r = Algorithms.ComputeEaster(year);
        Assert.Equal((ey, em, ed), (r.Year, r.Month, r.Day));
    }
}

public class AddDaysTests
{
    [Theory]
    [InlineData(2025, 1, 1, 10, 2025, 1, 11)]
    [InlineData(2025, 1, 1, -1, 2024, 12, 31)]
    [InlineData(2024, 2, 28, 1, 2024, 2, 29)]
    public void KnownOffsets(int y, int m, int d, int n, int ey, int em, int ed)
    {
        var r = Algorithms.AddDays(y, m, d, n);
        Assert.Equal((ey, em, ed), (r.Year, r.Month, r.Day));
    }
}

// ----------------------------------------------------------------------
// GregorianDate
// ----------------------------------------------------------------------

public class GregorianDateTests
{
    [Fact]
    public void BasicConstruction()
    {
        var g = new GregorianDate(2025, 1, 11);
        Assert.Equal((2025, 1, 11), (g.Year, g.Month, g.Day));
    }

    public static IEnumerable<object[]> InvalidVectors() =>
        TestVectors.V.Invalid_Gregorian_Dates.Select(d => new object[] { d.Year, d.Month, d.Day });

    [Theory]
    [MemberData(nameof(InvalidVectors))]
    public void RejectsInvalidDates(int y, int m, int d)
    {
        Assert.Throws<InvalidGregorianDateException>(() => new GregorianDate(y, m, d));
    }

    [Fact]
    public void ToDateOnly()
    {
        var g = new GregorianDate(2025, 1, 11);
        var dt = g.ToDateOnly();
        Assert.Equal(new DateOnly(2025, 1, 11), dt);
    }

    [Fact]
    public void FromDateOnly()
    {
        var dt = new DateOnly(2025, 1, 11);
        var g = GregorianDate.FromDateOnly(dt);
        Assert.Equal((2025, 1, 11), (g.Year, g.Month, g.Day));
    }

    [Fact]
    public void Equality()
    {
        var a = new GregorianDate(2025, 1, 11);
        var b = new GregorianDate(2025, 1, 11);
        Assert.Equal(a, b);
        Assert.Equal(a.GetHashCode(), b.GetHashCode());
    }
}

public class CopticDateTests
{
    [Fact]
    public void BasicConstruction()
    {
        var c = new CopticDate(1741, 5, 3);
        Assert.Equal((1741, 5, 3), (c.Year, c.Month, c.Day));
    }

    public static IEnumerable<object[]> InvalidVectors() =>
        TestVectors.V.Invalid_Coptic_Dates.Select(d => new object[] { d.Year, d.Month, d.Day });

    [Theory]
    [MemberData(nameof(InvalidVectors))]
    public void RejectsInvalidDates(int y, int m, int d)
    {
        Assert.Throws<InvalidCopticDateException>(() => new CopticDate(y, m, d));
    }

    [Fact]
    public void ToGregorian()
    {
        var c = new CopticDate(1741, 5, 3);
        var g = c.ToGregorian();
        Assert.Equal((2025, 1, 11), (g.Year, g.Month, g.Day));
    }

    [Fact]
    public void GregorianDateToCopticInstance()
    {
        var g = new GregorianDate(2025, 1, 11);
        var c = g.ToCoptic();
        Assert.Equal((1741, 5, 3), (c.Year, c.Month, c.Day));
    }
}

// ----------------------------------------------------------------------
// FeastsData parity with core/feasts.json
// ----------------------------------------------------------------------

public class FeastsDataTests
{
    [Fact]
    public void MatchesCoreFeastsJson()
    {
        var path = Path.Combine(Paths.CoreDir, "feasts.json");
        var json = File.ReadAllText(path);
        using var doc = JsonDocument.Parse(json);
        var coreArr = doc.RootElement;

        Assert.Equal(coreArr.GetArrayLength(), FeastsData.Feasts.Count);

        int i = 0;
        foreach (var el in coreArr.EnumerateArray())
        {
            var f = FeastsData.Feasts[i];
            Assert.Equal(el.GetProperty("id").GetString(), f.Id);
            Assert.Equal(el.GetProperty("type").GetString(), f.Type);
            Assert.Equal(el.GetProperty("category").GetString(), f.Category);
            var names = el.GetProperty("names");
            Assert.Equal(names.GetProperty("en").GetString(), f.Names["en"]);
            Assert.Equal(names.GetProperty("ar").GetString(), f.Names["ar"]);
            i++;
        }
    }
}

public class FeastTests
{
    [Fact]
    public void BasicFieldsAndName()
    {
        var g = new GregorianDate(2025, 4, 20);
        var rec = FeastsData.FeastById("easter");
        var feast = new Feast(rec.Id, rec.Type, rec.Category, rec.Names, g);
        Assert.Equal("easter", feast.Id);
        Assert.Equal("moveable", feast.Type);
        Assert.Equal("major", feast.Category);
        Assert.Equal("Easter Sunday", feast.Name("en"));
        Assert.Equal("عيد القيامة المجيد", feast.Name("ar"));
    }

    [Fact]
    public void UnknownLocaleThrows()
    {
        var g = new GregorianDate(2025, 4, 20);
        var rec = FeastsData.FeastById("easter");
        var feast = new Feast(rec.Id, rec.Type, rec.Category, rec.Names, g);
        Assert.Throws<UnsupportedLocaleException>(() => feast.Name("fr"));
    }
}

public class CopticCalendarEasterDateTests
{
    public static IEnumerable<object[]> Vectors() =>
        TestVectors.V.Easter.Select(v => new object[]
        {
            v.Gregorian_Year, v.Date.Year, v.Date.Month, v.Date.Day,
        });

    [Theory]
    [MemberData(nameof(Vectors))]
    public void EasterMatchesVectors(int year, int ey, int em, int ed)
    {
        var g = CopticCalendar.EasterDate(year);
        Assert.Equal((ey, em, ed), (g.Year, g.Month, g.Day));
    }
}

public class CopticCalendarMoveableFeastTests
{
    public static IEnumerable<object[]> Vectors() =>
        TestVectors.V.Moveable_Feasts.Select(v => new object[]
        {
            v.Feast_Id, v.Gregorian_Year, v.Date.Year, v.Date.Month, v.Date.Day,
        });

    [Theory]
    [MemberData(nameof(Vectors))]
    public void MoveableFeastVectors(string id, int year, int ey, int em, int ed)
    {
        var feast = CopticCalendar.MoveableFeast(id, year);
        Assert.Equal(id, feast.Id);
        Assert.Equal((ey, em, ed), (feast.GregorianDate.Year, feast.GregorianDate.Month, feast.GregorianDate.Day));
    }
}

public class CopticCalendarYearFeastsTests
{
    [Fact]
    public void NonEmptyAndSorted()
    {
        var feasts = CopticCalendar.YearFeasts(2025);
        Assert.NotEmpty(feasts);
        for (int i = 1; i < feasts.Count; i++)
        {
            var a = feasts[i - 1].GregorianDate;
            var b = feasts[i].GregorianDate;
            Assert.True(LessOrEqual(a, b), $"feasts not sorted at {i}: {a} > {b}");
        }
    }

    [Fact]
    public void IncludesEasterOn20250420()
    {
        var feasts = CopticCalendar.YearFeasts(2025);
        var easter = feasts.First(f => f.Id == "easter");
        var g = easter.GregorianDate;
        Assert.Equal((2025, 4, 20), (g.Year, g.Month, g.Day));
    }

    private static bool LessOrEqual(GregorianDate a, GregorianDate b)
    {
        if (a.Year != b.Year) return a.Year < b.Year;
        if (a.Month != b.Month) return a.Month < b.Month;
        return a.Day <= b.Day;
    }
}

// ----------------------------------------------------------------------
// CopticMonthsData parity with core/coptic_months.json
// ----------------------------------------------------------------------

public class CopticMonthsDataTests
{
    [Fact]
    public void MatchesCoreCopticMonthsJson()
    {
        var path = Path.Combine(Paths.CoreDir, "coptic_months.json");
        var json = File.ReadAllText(path);
        using var doc = JsonDocument.Parse(json);
        var core = doc.RootElement;

        Assert.Equal(core.GetArrayLength(), CopticMonthsData.Months.Count);
        for (int i = 0; i < CopticMonthsData.Months.Count; i++)
        {
            var m = CopticMonthsData.Months[i];
            var refEl = core[i];
            Assert.Equal(refEl.GetProperty("month").GetInt32(), m.Month);
            var refNames = refEl.GetProperty("names");
            Assert.Equal(refNames.GetProperty("en").GetString(), m.Names["en"]);
            Assert.Equal(refNames.GetProperty("ar").GetString(), m.Names["ar"]);
        }
    }
}

// ----------------------------------------------------------------------
// CopticCalendar.MonthName
// ----------------------------------------------------------------------

public class CopticCalendarMonthNameTests
{
    public static IEnumerable<object[]> Vectors() =>
        TestVectors.V.Coptic_Month_Names.Select(v => new object[]
        {
            v.Month, v.Locale, v.Name,
        });

    [Theory]
    [MemberData(nameof(Vectors))]
    public void Vectors_Match(int month, string locale, string name)
    {
        Assert.Equal(name, CopticCalendar.MonthName(month, locale));
    }

    public static IEnumerable<object[]> InvalidMonths() =>
        TestVectors.V.Invalid_Coptic_Months_For_Name.Select(m => new object[] { m });

    [Theory]
    [MemberData(nameof(InvalidMonths))]
    public void RejectsInvalidMonth(int month)
    {
        Assert.Throws<InvalidCopticMonthException>(
            () => CopticCalendar.MonthName(month, "en"));
    }

    public static IEnumerable<object[]> InvalidLocales() =>
        TestVectors.V.Invalid_Coptic_Month_Locales.Select(v => new object[] { v.Month, v.Locale });

    [Theory]
    [MemberData(nameof(InvalidLocales))]
    public void RejectsUnsupportedLocale(int month, string locale)
    {
        Assert.Throws<UnsupportedLocaleException>(
            () => CopticCalendar.MonthName(month, locale));
    }
}
